agent/php_newrelic.h

```diff
@@ -409,7 +409,10 @@ size_t drupal_module_invoke_all_hook_len; /* The length of the current Drupal
                                              hook */
 size_t drupal_http_request_depth; /* The current depth of drupal_http_request()
                                      calls */
-
+#if ZEND_MODULE_API_NO >= ZEND_8_0_X_API_NO \
+    && !defined OVERWRITE_ZEND_EXECUTE_DATA
+nr_segment_t* drupal_http_request_segment;
```

Comment:
Does this need to be initialized anywhere or is it assumed to be set to zero by the compiler because of the global scope?

Reply:
It doesn't need to be initialized to anything other than `NULL` which is handled by the complier because of global scope. The segment assigned to `drupal_http_request_segment` is created [here](https://github.com/misiekn/newrelic-php-agent/blob/oapi/drupal_http_request/agent/fw_drupal.c#L243) and unassigned [here](https://github.com/misiekn/newrelic-php-agent/blob/oapi/drupal_http_request/agent/fw_drupal.c#L291).

Reply:
Should we doublecheck that this got freed in rshutdown?
Maybe something here: https://github.com/newrelic/newrelic-php-agent/blob/main/agent/php_rshutdown.c#L92
Since the segment start and stop are in different functions now we have to be prepared for something happening in between.

Reply:
`drupal_http_request_segment` points to memory chunk allocated by slab allocator (`nr_segment_start` calls [nr_txn_allocate_segment](https://github.com/newrelic/newrelic-php-agent/blob/main/axiom/nr_segment.c#L179) which in turn calls [nr_slab_next](https://github.com/newrelic/newrelic-php-agent/blob/main/axiom/nr_txn.h#L1138)). It is guaranteed to be freed by slab allocator [here](https://github.com/newrelic/newrelic-php-agent/blob/main/axiom/nr_txn.c#L1237) (as part of  `nr_txn_destroy_fields` which is called from within `nr_txn_destroy` which is called by `nr_php_txn_end` called in `nr_php_post_deactivate`). I can add setting `drupal_http_request_segment` to NULL in `nr_php_post_deactivate`.

Reply:
@mfulb @zsistla I tried to address these concerns with c30a987486a5bc2ea8e6c28fd470be5d593874d7 and e4128608b6df37f1892e6e5ef505fcaff8c0d1f4. Is this acceptable?

Reply:
The "exception path" in e412860 would be if the segment was not kicked from the segment minheap and persists to the end of the request?

Reply:
> The "exception path" in https://github.com/newrelic/newrelic-php-agent/commit/e4128608b6df37f1892e6e5ef505fcaff8c0d1f4 would be if the segment was not kicked from the segment minheap and persists to the end of the request?

The 'exception path' covers anything that does not fall under the 'happy path' category, i.e. segment ended in the _after_ callback. The change in e4128608b6df37f1892e6e5ef505fcaff8c0d1f4 does not ensure that the memory pointed by `drupal_http_request_segment` is freed - this is guaranteed to be freed by slab allocator [here](https://github.com/newrelic/newrelic-php-agent/blob/main/axiom/nr_txn.c#L1237). The change in e4128608b6df37f1892e6e5ef505fcaff8c0d1f4 only ensures that at the end of the request, the `drupal_http_request_segment` pointer will no longer point to anything. In my tests `drupal_http_request_segment` pointer had non-NULL value regardless of the outcome of `drupal_http_request` (successful execution or execution ended by exception), and regardless if external segment was or was not moved to min-max heap (moving the segment to the min-max heap should invalidate the pointer pointing to the segment but it is not happening). This change is last resort failsafe to ensure accessing `drupal_http_request_segment` outside the scope of `nr_drupal_http_request_before` and `nr_drupal_http_request_after` callbacks will not cause unexpected memory corruption.

FYI: I tested `drupal_http_request` execution with and without exception as well as nested calls to `drupal_http_request` caused by redirect.